File: fce-extension/fce-sign/python/app/handlers.py
"""Handler functions for the KEY extension operations."""

from __future__ import annotations

import base64
import json
import logging
from typing import Any, Optional
from urllib.request import urlopen, Request
from urllib.error import URLError, HTTPError

from base.types import Framework
from .config import VERSION, OP_TYPE_KEY, OP_COMMAND_UPDATE, OP_COMMAND_SIGN
from .abi import abi_encode_two
from .crypto import sign_ecdsa, parse_private_key
from base.encoding import hex_to_bytes, bytes_to_hex

logger = logging.getLogger(__name__)

# Mutable state — the framework serializes all handler calls.
_private_key: Optional[bytes] = None
_sign_port: str = "9090"


def set_sign_port(port: str) -> None:
    """Set the sign port for communicating with the TEE node."""
    global _sign_port
    _sign_port = port


def register(framework: Framework) -> None:
    """Register the KEY handlers with the framework."""
    framework.handle(OP_TYPE_KEY, OP_COMMAND_UPDATE, handle_key_update)
    framework.handle(OP_TYPE_KEY, OP_COMMAND_SIGN, handle_key_sign)


def report_state() -> Any:
    """Return a JSON-serializable snapshot of the current state."""
    return {
        "hasKey": _private_key is not None,
        "version": VERSION,
    }


def handle_key_update(msg: str) -> tuple[Optional[str], int, Optional[str]]:
    """Decrypt the original message using the TEE node's key, store as private key."""
    global _private_key

    if not msg:
        return None, 0, "originalMessage is empty"

    try:
        ciphertext = hex_to_bytes(msg)
    except Exception as e:
        return None, 0, f"invalid hex in originalMessage: {e}"

    try:
        key_bytes = _decrypt_via_node(ciphertext)
    except Exception as e:
        return None, 0, f"decryption failed: {e}"

    try:
        validated_key = parse_private_key(key_bytes)
    except Exception as e:
        return None, 0, f"invalid private key: {e}"

    _private_key = validated_key
    logger.info("private key updated")
    return None, 1, None


def handle_key_sign(msg: str) -> tuple[Optional[str], int, Optional[str]]:
    """Sign the original message with the stored private key."""
    if _private_key is None:
        return None, 0, "no private key stored"

    if not msg:
        return None, 0, "originalMessage is empty"

    try:
        msg_bytes = hex_to_bytes(msg)
    except Exception as e:
        return None, 0, f"invalid hex in originalMessage: {e}"

    try:
        sig = sign_ecdsa(_private_key, msg_bytes)
    except Exception as e:
        return None, 0, f"signing failed: {e}"

    try:
        encoded = abi_encode_two(msg_bytes, sig)
    except Exception as e:
        return None, 0, f"ABI encoding failed: {e}"

    data_hex = bytes_to_hex(encoded)
    return data_hex, 1, None


def _decrypt_via_node(ciphertext: bytes) -> bytes:
    """Call the TEE node's /decrypt endpoint.

    Sends ciphertext as base64-encoded bytes (matching Go's []byte JSON marshaling).
    Returns the decrypted plaintext bytes.
    """
    url = f"http://localhost:{_sign_port}/decrypt"
    req_body = json.dumps({
        "encryptedMessage": base64.b64encode(ciphertext).decode("ascii"),
    }).encode("utf-8")

    req = Request(url, data=req_body, headers={"Content-Type": "application/json"})
    try:
        with urlopen(req) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            return base64.b64decode(data["decryptedMessage"])
    except HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"node returned {e.code}: {body}") from e
    except URLError as e:
        raise RuntimeError(f"request error: {e}") from e

File: fce-extension/typescript/src/app/handlers.ts
/** Handler functions for the KEY extension operations. */

import http from 'node:http';
import { Framework } from '../base/types.js';
import {
  VERSION,
  OP_TYPE_KEY,
  OP_COMMAND_UPDATE,
  OP_COMMAND_SIGN,
} from './config.js';
import { abiEncodeTwo } from './abi.js';
import { signECDSA, parsePrivateKey } from './crypto.js';
import { hexToBytes, bytesToHex } from '../base/encoding.js';

/** Mutable state — the framework serializes all handler calls. */
let privateKey: Uint8Array | null = null;
let signPort = '9090';

/** Set the sign port for communicating with the TEE node. */
export function setSignPort(port: string): void {
  signPort = port;
}

/** Register the KEY handlers with the framework. */
export function register(framework: Framework): void {
  framework.handle(OP_TYPE_KEY, OP_COMMAND_UPDATE, handleKeyUpdate);
  framework.handle(OP_TYPE_KEY, OP_COMMAND_SIGN, handleKeySign);
}

/** Return a JSON-serializable snapshot of the current state. */
export function reportState(): unknown {
  return {
    hasKey: privateKey !== null,
    version: VERSION,
  };
}

/** Reset state (for testing). */
export function resetState(): void {
  privateKey = null;
}

async function handleKeyUpdate(
  msg: string,
): Promise<[string | null, number, string | null]> {
  if (!msg) {
    return [null, 0, 'originalMessage is empty'];
  }

  let ciphertext: Uint8Array;
  try {
    ciphertext = hexToBytes(msg);
  } catch (e) {
    return [null, 0, `invalid hex in originalMessage: ${e}`];
  }

  let keyBytes: Uint8Array;
  try {
    keyBytes = await decryptViaNode(ciphertext);
  } catch (e) {
    return [null, 0, `decryption failed: ${e}`];
  }

  let validatedKey: Uint8Array;
  try {
    validatedKey = parsePrivateKey(keyBytes);
  } catch (e) {
    return [null, 0, `invalid private key: ${e}`];
  }

  privateKey = validatedKey;
  console.log('private key updated');
  return [null, 1, null];
}

async function handleKeySign(
  msg: string,
): Promise<[string | null, number, string | null]> {
  if (privateKey === null) {
    return [null, 0, 'no private key stored'];
  }

  if (!msg) {
    return [null, 0, 'originalMessage is empty'];
  }

  let msgBytes: Uint8Array;
  try {
    msgBytes = hexToBytes(msg);
  } catch (e) {
    return [null, 0, `invalid hex in originalMessage: ${e}`];
  }

  let sig: Uint8Array;
  try {
    sig = signECDSA(privateKey, msgBytes);
  } catch (e) {
    return [null, 0, `signing failed: ${e}`];
  }

  let encoded: Uint8Array;
  try {
    encoded = abiEncodeTwo(msgBytes, sig);
  } catch (e) {
    return [null, 0, `ABI encoding failed: ${e}`];
  }

  const dataHex = bytesToHex(encoded);
  return [dataHex, 1, null];
}

/**
 * Call the TEE node's /decrypt endpoint.
 * Sends ciphertext as base64-encoded bytes (matching Go's []byte JSON marshaling).
 * Returns the decrypted plaintext bytes.
 */
function decryptViaNode(ciphertext: Uint8Array): Promise<Uint8Array> {
  return new Promise((resolve, reject) => {
    const url = `http://localhost:${signPort}/decrypt`;
    const body = JSON.stringify({
      encryptedMessage: Buffer.from(ciphertext).toString('base64'),
    });

    const req = http.request(
      url,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(body),
        },
      },
      (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (chunk: Buffer) => chunks.push(chunk));
        res.on('end', () => {
          const data = Buffer.concat(chunks).toString('utf-8');
          if (res.statusCode !== 200) {
            reject(new Error(`node returned ${res.statusCode}: ${data}`));
            return;
          }
          try {
            const parsed = JSON.parse(data);
            resolve(
              new Uint8Array(Buffer.from(parsed.decryptedMessage, 'base64')),
            );
          } catch (e) {
            reject(new Error(`decode response: ${e}`));
          }
        });
      },
    );

    req.on('error', (e) => reject(new Error(`request error: ${e.message}`)));
    req.write(body);
    req.end();
  });
}
